src/uct/ib/base/ib_iface.c

```diff
@@ -247,87 +251,109 @@ void uct_ib_iface_release_desc(uct_recv_desc_t *self, void *desc)
     ucs_mpool_put_inline(ib_desc);
 }
 
-size_t uct_ib_address_size(const union ibv_gid *gid, unsigned pack_flags)
+size_t uct_ib_address_size(const uct_ib_address_pack_params_t *params)
 {
     size_t size = sizeof(uct_ib_address_t);
 
-    if (pack_flags & UCT_IB_ADDRESS_PACK_FLAG_ETH) {
+    if (params->flags & UCT_IB_ADDRESS_PACK_FLAG_ETH) {
         /* Ethernet: address contains only raw GID */
-        return size + sizeof(union ibv_gid);
-    }
+        size += sizeof(union ibv_gid);
+    } else {
+        /* InfiniBand: address always contains LID */
+        size += sizeof(uint16_t); /* lid */
+
+        if (params->flags & UCT_IB_ADDRESS_PACK_FLAG_INTERFACE_ID) {
+            /* Add GUID */
+            UCS_STATIC_ASSERT(sizeof(params->gid->global.interface_id) == sizeof(uint64_t));
+            size += sizeof(uint64_t);
+        }
 
-    /* InfiniBand: address always contains LID */
-    size += sizeof(uint16_t); /* lid */
+        if (params->flags & UCT_IB_ADDRESS_PACK_FLAG_SUBNET_PREFIX) {
+            if ((params->gid->global.subnet_prefix & UCT_IB_SITE_LOCAL_MASK) ==
+                                                     UCT_IB_SITE_LOCAL_PREFIX) {
+                /* 16-bit subnet prefix */
+                size += sizeof(uint16_t);
+            } else if (params->gid->global.subnet_prefix != UCT_IB_LINK_LOCAL_PREFIX) {
+                /* 64-bit subnet prefix */
+                size += sizeof(uint64_t);
+            }
+            /* Note: if subnet prefix is LINK_LOCAL, no need to pack it because
+             * it's a well-known value defined by IB specification.
+             */
+        }
+    }
 
-    if (pack_flags & UCT_IB_ADDRESS_PACK_FLAG_INTERFACE_ID) {
-        /* Add GUID */
-        UCS_STATIC_ASSERT(sizeof(gid->global.interface_id) == sizeof(uint64_t));
-        size += sizeof(uint64_t);
+    if (params->flags & UCT_IB_ADDRESS_PACK_FLAG_GID_INDEX) {
```

Comment:
pack ordering is defined by `UCT_IB_ADDRESS_PACK_xx` not `UCT_IB_ADDRESS_FLAG_xx`, do you suggest to reorder the first enum as well? pls note this may cause incompatible address format, also will affect all cases, not only new ones. IMO it's more risky.